x-pack/plugins/reporting/server/plugin.ts

```diff
@@ -20,9 +20,26 @@ export class ReportingPlugin
   private logger: LevelLogger;
   private reportingCore?: ReportingCore;
 
+  private setupDone: (val?: unknown) => void = () => {};
+  private startDone: (val?: unknown) => void = () => {};
+  private setupPromise: Promise<unknown>;
+  private startPromise: Promise<unknown>;
+
   constructor(context: PluginInitializerContext<ReportingConfigType>) {
     this.logger = new LevelLogger(context.logger.get());
     this.initializerContext = context;
+
+    // Setup some promises for modules that need to await setup/start
+    this.setupPromise = new Promise((r) => (this.setupDone = r));
+    this.startPromise = new Promise((r) => (this.startDone = r));
+  }
+
+  public async setupReady() {
+    return this.setupPromise;
+  }
+
+  public async startReady() {
+    return this.startPromise;
```

Comment:
I was looking at the `setupReady` and `startReady` methods, and thought, reportingCore can have methods like that using the internal state it already has. I think these break out of ReportingCore's centralized state.

```
  public async setupReady() {
    return this.reportingCore.pluginHasSetUp(); // need to implement
  }

  public async startReady() {
    return this.reportingCore.pluginHasStarted(); // this already exists
  }
```

Reply:
Hm, `this.reportingCore` could be undefined because it depends on config..

Reply:
++ agreed, will move in a follow up PR

Reply:
nm, I took a look at shoving the implementation details into `core`, but because core setup is async there's still a lot of detail in `plugin`, so I think it makes sense to keep it this way until we can refactor things out in a more `sync`-ish way.